backend/cfg/cfg_prologue.ml

```diff
@@ -58,7 +86,123 @@ let add_prologue_if_required : Cfg_with_layout.t -> Cfg_with_layout.t =
           ()));
   cfg_with_layout
 
+let fun_name = ref ""
+
+module Validator = struct
+  type state =
+    | No_prologue_on_stack
+    | Prologue_on_stack
+
+  (* This is necessary to make a set, but the ordering of elements is
+     arbitrary. *)
+  let state_compare left right =
+    match left, right with
+    | No_prologue_on_stack, No_prologue_on_stack -> 0
+    | No_prologue_on_stack, Prologue_on_stack -> 1
+    | Prologue_on_stack, No_prologue_on_stack -> -1
+    | Prologue_on_stack, Prologue_on_stack -> 0
+
+  module State_set = Set.Make (struct
+    type t = state
+
+    let compare = state_compare
+  end)
+
+  (* The validator domain represents the set of possible states at an
+     instruction (i.e. a state {Prologue_on_stack, No_prologue_on_stack} means
+     that depending on the execution path used to get to that block/instruction,
+     we can either have a prologue on the stack or not).
+
+     Non-singleton states are allowed in cases where there is no Prologue,
+     Epilogue nor any instructions which require a prologue (this happens e.g.
+     when two [raise] terminators reach the same handler, but one is before the
+     prologue, and the other is after the prologue). *)
```

Comment:
Maybe the example should be slightly
expended to make it clearer why this
is not immediately an error.